File: data/src/notFoss/java/org/cryptomator/data/cloud/googledrive/GoogleDriveCloudContentRepository.java
package org.cryptomator.data.cloud.googledrive;

import android.content.Context;

import com.google.api.client.googleapis.extensions.android.gms.auth.UserRecoverableAuthIOException;
import com.google.api.client.googleapis.json.GoogleJsonResponseException;
import com.google.api.client.http.HttpStatusCodes;

import org.cryptomator.data.cloud.InterceptingCloudContentRepository;
import org.cryptomator.domain.CloudNode;
import org.cryptomator.domain.GoogleDriveCloud;
import org.cryptomator.domain.exception.BackendException;
import org.cryptomator.domain.exception.FatalBackendException;
import org.cryptomator.domain.exception.NetworkConnectionException;
import org.cryptomator.domain.exception.NoSuchCloudFileException;
import org.cryptomator.domain.exception.authentication.UserRecoverableAuthenticationException;
import org.cryptomator.domain.repository.CloudContentRepository;
import org.cryptomator.domain.usecases.ProgressAware;
import org.cryptomator.domain.usecases.cloud.DataSource;
import org.cryptomator.domain.usecases.cloud.DownloadState;
import org.cryptomator.domain.usecases.cloud.UploadState;
import org.cryptomator.util.ExceptionUtil;
import org.cryptomator.util.Optional;

import java.io.File;
import java.io.IOException;
import java.io.OutputStream;
import java.net.SocketTimeoutException;
import java.util.List;

import static org.cryptomator.util.ExceptionUtil.contains;
import static org.cryptomator.util.ExceptionUtil.extract;

class GoogleDriveCloudContentRepository extends InterceptingCloudContentRepository<GoogleDriveCloud, GoogleDriveNode, GoogleDriveFolder, GoogleDriveFile> {

	private final GoogleDriveCloud cloud;

	GoogleDriveCloudContentRepository(Context context, GoogleDriveCloud cloud, GoogleDriveIdCache idCache) {
		super(new Intercepted(context, cloud, idCache));
		this.cloud = cloud;
	}

	@Override
	protected void throwWrappedIfRequired(Exception e) throws BackendException {
		throwConnectionErrorIfRequired(e);
		throwUserRecoverableAuthenticationExceptionIfRequired(e);
		throwNoSuchCloudFileExceptionIfRequired(e);
	}

	private void throwUserRecoverableAuthenticationExceptionIfRequired(Exception e) {
		Optional<UserRecoverableAuthIOException> userRecoverableAuthIOException = extract(e, UserRecoverableAuthIOException.class);
		if (userRecoverableAuthIOException.isPresent()) {
			throw new UserRecoverableAuthenticationException(cloud, userRecoverableAuthIOException.get().getIntent());
		}
	}

	private void throwConnectionErrorIfRequired(Exception e) throws NetworkConnectionException {
		if (contains(e, SocketTimeoutException.class) || contains(e, IOException.class, ExceptionUtil.thatHasMessage("NetworkError"))) {
			throw new NetworkConnectionException(e);
		}
	}

	private void throwNoSuchCloudFileExceptionIfRequired(Exception e) throws NoSuchCloudFileException {
		if (contains(e, GoogleJsonResponseException.class)) {
			if (extract(e, GoogleJsonResponseException.class).get().getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
				throw new NoSuchCloudFileException();
			}
		}
	}

	private static class Intercepted implements CloudContentRepository<GoogleDriveCloud, GoogleDriveNode, GoogleDriveFolder, GoogleDriveFile> {

		private final GoogleDriveImpl impl;

		public Intercepted(Context context, GoogleDriveCloud cloud, GoogleDriveIdCache idCache) {
			this.impl = new GoogleDriveImpl(context, cloud, idCache);
		}

		@Override
		public GoogleDriveFolder root(GoogleDriveCloud cloud) throws BackendException {
			return impl.root();
		}

		@Override
		public GoogleDriveFolder resolve(GoogleDriveCloud cloud, String path) throws BackendException {
			try {
				return impl.resolve(path);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public GoogleDriveFile file(GoogleDriveFolder parent, String name) throws BackendException {
			try {
				return impl.file(parent, name);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public GoogleDriveFile file(GoogleDriveFolder parent, String name, Optional<Long> size) throws BackendException {
			try {
				return impl.file(parent, name, size);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public GoogleDriveFolder folder(GoogleDriveFolder parent, String name) throws BackendException {
			try {
				return impl.folder(parent, name);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public boolean exists(GoogleDriveNode node) throws BackendException {
			try {
				return impl.exists(node);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public List<CloudNode> list(GoogleDriveFolder folder) throws BackendException {
			try {
				return impl.list(folder);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public GoogleDriveFolder create(GoogleDriveFolder folder) throws BackendException {
			try {
				return impl.create(folder);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public GoogleDriveFolder move(GoogleDriveFolder source, GoogleDriveFolder target) throws BackendException {
			try {
				if (source.getDriveId() == null) {
					throw new NoSuchCloudFileException(source.getName());
				}
				return (GoogleDriveFolder) impl.move(source, target);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public GoogleDriveFile move(GoogleDriveFile source, GoogleDriveFile target) throws BackendException {
			try {
				return (GoogleDriveFile) impl.move(source, target);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public GoogleDriveFile write(GoogleDriveFile file, DataSource data, ProgressAware<UploadState> progressAware, boolean replace, long size) throws BackendException {
			try {
				return impl.write(file, data, progressAware, replace, size);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public void read(GoogleDriveFile file, Optional<File> encryptedTmpFile, OutputStream data, ProgressAware<DownloadState> progressAware) throws BackendException {
			try {
				if (file.getDriveId() == null) {
					throw new NoSuchCloudFileException(file.getName());
				}
				impl.read(file, encryptedTmpFile, data, progressAware);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public void delete(GoogleDriveNode node) throws BackendException {
			try {
				impl.delete(node);
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public String checkAuthenticationAndRetrieveCurrentAccount(GoogleDriveCloud cloud) throws BackendException {
			try {
				return impl.currentAccount();
			} catch (IOException e) {
				throw new FatalBackendException(e);
			}
		}

		@Override
		public void logout(GoogleDriveCloud cloud) throws BackendException {
			// empty
		}
	}

}

File: data/src/apkStorePlaystore/java/org/cryptomator/data/cloud/googledrive/GoogleDriveCloudContentRepository.kt
package org.cryptomator.data.cloud.googledrive

import android.content.Context
import com.google.api.client.googleapis.extensions.android.gms.auth.UserRecoverableAuthIOException
import com.google.api.client.googleapis.json.GoogleJsonResponseException
import com.google.api.client.http.HttpStatusCodes
import org.cryptomator.data.cloud.InterceptingCloudContentRepository
import org.cryptomator.domain.GoogleDriveCloud
import org.cryptomator.domain.exception.BackendException
import org.cryptomator.domain.exception.FatalBackendException
import org.cryptomator.domain.exception.NetworkConnectionException
import org.cryptomator.domain.exception.NoSuchCloudFileException
import org.cryptomator.domain.exception.authentication.UserRecoverableAuthenticationException
import org.cryptomator.domain.repository.CloudContentRepository
import org.cryptomator.domain.usecases.ProgressAware
import org.cryptomator.domain.usecases.cloud.DataSource
import org.cryptomator.domain.usecases.cloud.DownloadState
import org.cryptomator.domain.usecases.cloud.UploadState
import org.cryptomator.util.ExceptionUtil
import java.io.File
import java.io.IOException
import java.io.OutputStream
import java.net.SocketTimeoutException

internal class GoogleDriveCloudContentRepository(context: Context, private val cloud: GoogleDriveCloud, idCache: GoogleDriveIdCache) :
	InterceptingCloudContentRepository<GoogleDriveCloud, GoogleDriveNode, GoogleDriveFolder, GoogleDriveFile>(Intercepted(context, cloud, idCache)) {

	@Throws(BackendException::class)
	override fun throwWrappedIfRequired(e: Exception) {
		throwConnectionErrorIfRequired(e)
		throwUserRecoverableAuthenticationExceptionIfRequired(e)
		throwNoSuchCloudFileExceptionIfRequired(e)
	}

	private fun throwUserRecoverableAuthenticationExceptionIfRequired(e: Exception) {
		val userRecoverableAuthIOException = ExceptionUtil.extract(e, UserRecoverableAuthIOException::class.java)
		if (userRecoverableAuthIOException.isPresent) {
			throw UserRecoverableAuthenticationException(cloud, userRecoverableAuthIOException.get().intent)
		}
	}

	@Throws(NetworkConnectionException::class)
	private fun throwConnectionErrorIfRequired(e: Exception) {
		if (ExceptionUtil.contains(e, SocketTimeoutException::class.java) || ExceptionUtil.contains(e, IOException::class.java, ExceptionUtil.thatHasMessage("NetworkError"))) {
			throw NetworkConnectionException(e)
		}
	}

	@Throws(NoSuchCloudFileException::class)
	private fun throwNoSuchCloudFileExceptionIfRequired(e: Exception) {
		if (ExceptionUtil.contains(e, GoogleJsonResponseException::class.java)) {
			if (ExceptionUtil.extract(e, GoogleJsonResponseException::class.java).get().statusCode == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
				throw NoSuchCloudFileException()
			}
		}
	}

	private class Intercepted(context: Context, cloud: GoogleDriveCloud, idCache: GoogleDriveIdCache) : CloudContentRepository<GoogleDriveCloud, GoogleDriveNode, GoogleDriveFolder, GoogleDriveFile> {

		private val impl: GoogleDriveImpl = GoogleDriveImpl(context, cloud, idCache)

		@Throws(BackendException::class)
		override fun root(cloud: GoogleDriveCloud): GoogleDriveFolder {
			return impl.root()
		}

		override fun resolve(cloud: GoogleDriveCloud, path: String): GoogleDriveFolder {
			return try {
				impl.resolve(path)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun file(parent: GoogleDriveFolder, name: String): GoogleDriveFile {
			return try {
				impl.file(parent, name, null)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun file(parent: GoogleDriveFolder, name: String, size: Long?): GoogleDriveFile {
			return try {
				impl.file(parent, name, size)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun folder(parent: GoogleDriveFolder, name: String): GoogleDriveFolder {
			return try {
				impl.folder(parent, name)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun exists(node: GoogleDriveNode): Boolean {
			return try {
				impl.exists(node)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun list(folder: GoogleDriveFolder): List<GoogleDriveNode> {
			return try {
				impl.list(folder)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun create(folder: GoogleDriveFolder): GoogleDriveFolder {
			return try {
				impl.create(folder)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun move(source: GoogleDriveFolder, target: GoogleDriveFolder): GoogleDriveFolder {
			return try {
				if (source.driveId == null) {
					throw NoSuchCloudFileException(source.name)
				}
				impl.move(source, target) as GoogleDriveFolder
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun move(source: GoogleDriveFile, target: GoogleDriveFile): GoogleDriveFile {
			return try {
				impl.move(source, target) as GoogleDriveFile
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun write(file: GoogleDriveFile, data: DataSource, progressAware: ProgressAware<UploadState>, replace: Boolean, size: Long): GoogleDriveFile {
			return try {
				impl.write(file, data, progressAware, replace, size)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun read(file: GoogleDriveFile, encryptedTmpFile: File?, data: OutputStream, progressAware: ProgressAware<DownloadState>) {
			try {
				if (file.driveId == null) {
					throw NoSuchCloudFileException(file.name)
				}
				impl.read(file, encryptedTmpFile, data, progressAware)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun delete(node: GoogleDriveNode) {
			try {
				impl.delete(node)
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun checkAuthenticationAndRetrieveCurrentAccount(cloud: GoogleDriveCloud): String {
			return try {
				impl.currentAccount()
			} catch (e: IOException) {
				throw FatalBackendException(e)
			}
		}

		@Throws(BackendException::class)
		override fun logout(cloud: GoogleDriveCloud) {
			impl.logout()
		}

	}
}
